drivers/vulkan/rendering_device_vulkan.cpp

```diff
@@ -8765,6 +8800,7 @@ void RenderingDeviceVulkan::submit() {
 
 	_finalize_command_bufers();
 
+	const uint32_t frame = context->get_frame_index();
 	VkCommandBuffer command_buffers[2] = { frames[frame].setup_command_buffer, frames[frame].draw_command_buffer };
```

Comment:
The design of the LocalRenderingDevice is essentially exposing a RenderingDevice without a swapchain or concept of a frame. Instead of swapping buffers, users are expected to call ``submit()`` and ``sync()`` and convenient times so that they can schedule work as they want to. 

I think having a global ``frame`` index can make sense for multiple windows and such, but it doesn't really make sense for the local rendering device.

Reply:
> Consider the following example:

No. That must never happen.

If the frame index is about to be increased, we **must** first submit and a VkFence issued so we can wait on it.

At least when it comes to what VulkanContext considers a 'frame'. If the higher level interface wants to render two frames and submit them to VulkanContext as a single 'frame' that's alright.

Reply:
> > Consider the following example:
> 
> No. That must never happen.
> 
> If the frame index is about to be increased, we **must** first submit and a VkFence issued so we can wait on it.
> 
> At least when it comes to what VulkanContext considers a 'frame'. If the higher level interface wants to render two frames and submit them to VulkanContext as a single 'frame' that's alright.

I think you are misunderstanding what a local rendering device is. The entire point of it is to have a place to submit commands over multiple frames. For work that is okay to be run in the background. When the user is done recording, they call submit, then they cannot record or submit until they call sync. 

Logically, this doesn't map onto the concept of frames. You have a single queue, a single command_buffer. You submit when you are done recording, and you sync before you start recording again. There is no need for multiple frames, and indeed, having multiple frames makes things way more complicated as you need to start to manage what work was done in which frame. 

The idea behind the LocalRenderingDevice is that when users need per-frame stuff, they just use the global rendering device, if they are okay having something run in the background over multiple frames, they can use the localrenderingdevice

Reply:
If it helps, 'local rendering device' is a headless thing.